src/arns.lua

```diff
@@ -355,28 +355,65 @@ function arns.getProcessIdForRecord(name)
 	return record ~= nil and record.processId or nil
 end
 
+local function getActiveArNSNamesBetweenTimestampsFromRecords(records, startTimestamp, endTimestamp)
+	local activeNames = {}
+	local namesInGracePeriod = {}
+
+	for name, record in pairs(records) do
+		if record.type == "permabuy" then
+			table.insert(activeNames, name)
+		elseif
+			record.type == "lease"
+			and record.endTimestamp
+			and record.startTimestamp
+			and record.startTimestamp <= startTimestamp
+		then
+			if record.endTimestamp >= endTimestamp then
+				table.insert(activeNames, name)
+			elseif record.endTimestamp + constants.gracePeriodMs >= endTimestamp then
+				table.insert(namesInGracePeriod, name)
+			end
+		end
+	end
+	return { activeNames = activeNames, namesInGracePeriod = namesInGracePeriod }
+end
+
 --- Gets the active ARNS names between two timestamps
 --- @param startTimestamp number The start timestamp
 --- @param endTimestamp number The end timestamp
 --- @return table<string> The active ARNS names between the two timestamps
 function arns.getActiveArNSNamesBetweenTimestamps(startTimestamp, endTimestamp)
-	local records = arns.getRecords()
-	local activeNames = {}
-	for name, record in pairs(records) do
-		if
-			record.type == "permabuy"
-			or (
-				record.type == "lease"
-				and record.endTimestamp
-				and record.startTimestamp
-				and record.startTimestamp <= startTimestamp
-				and record.endTimestamp >= endTimestamp
-			)
-		then
-			table.insert(activeNames, name)
+	return getActiveArNSNamesBetweenTimestampsFromRecords(arns.getRecords(), startTimestamp, endTimestamp).activeNames
```

Comment:
nit - use `getUnsafeRecords` to avoid the copying

Reply:
this was an existing path grabbing the deepCopy implementation used by computePrescribedNamesForEpoch. Is that okay to use unsafe here?
<img width="1163" alt="image" src="https://github.com/user-attachments/assets/28704008-9052-466d-99d5-acdc409b809b" />

Reply:
reasoning through it. I guess as long as we're not modifying the records themsleves then re-writing them its fine to use unsafe. here we're just harvesting data for creating the epoch

Reply:
3798a5f